test/extensions/filters/http/ext_authz/ext_authz_test.cc

```diff
@@ -375,6 +375,55 @@ TEST_P(HttpExtAuthzFilterParamTest, DeniedResponseWith403) {
       cm_.thread_local_cluster_.cluster_.info_->stats_store_.counter("upstream_rq_403").value());
 }
 
+// Verify that authz response memory is not used after free.
+TEST_P(HttpExtAuthzFilterParamTest, DestroyResponseBeforeSendLocalReply) {
+  InSequence s;
+
+  Filters::Common::ExtAuthz::Response response{};
+  response.status = Filters::Common::ExtAuthz::CheckStatus::Denied;
+  response.status_code = Http::Code::Forbidden;
+  response.body = std::string{"foo"};
+  response.headers_to_add = Http::HeaderVector{{Http::LowerCaseString{"foo"}, "bar"},
+                                               {Http::LowerCaseString{"bar"}, "foo"}};
+  Filters::Common::ExtAuthz::ResponsePtr response_ptr =
+      std::make_unique<Filters::Common::ExtAuthz::Response>(response);
+
+  ON_CALL(filter_callbacks_, connection()).WillByDefault(Return(&connection_));
+  EXPECT_CALL(connection_, remoteAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(connection_, localAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(*client_, check(_, _, _))
+      .WillOnce(
+          WithArgs<0>(Invoke([&](Filters::Common::ExtAuthz::RequestCallbacks& callbacks) -> void {
+            request_callbacks_ = &callbacks;
+          })));
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
+            filter_->decodeHeaders(request_headers_, false));
+
+  Http::TestHeaderMapImpl response_headers{{":status", "403"},
+                                           {"content-length", "3"},
+                                           {"content-type", "text/plain"},
+                                           {"foo", "bar"},
+                                           {"bar", "foo"}};
+  EXPECT_CALL(filter_callbacks_, encodeHeaders_(HeaderMapEqualRef(&response_headers), false))
+      .WillOnce(
+          Invoke([&response_ptr](Http::TestHeaderMapImpl, bool) -> void { response_ptr.reset(); }));
```

Comment:
what you want to do here is iterate though the headers and check them after you reset the response pointer. I believe this would have crashed w/ the old code under ASAN.

Reply:
I will do that. Thanks!